packages/swing-store/src/snapStore.js

```diff
@@ -217,63 +213,117 @@ export function makeSnapStore(
 
   /** @param {string} hash */
   function fullPathFromHash(hash) {
-    return resolve(root, baseNameFromHash(hash));
+    return pathResolve(root, baseNameFromHash(hash));
   }
 
   /** @type { Set<string> } */
   const toDelete = new Set();
 
   /**
+   * Creates a new gzipped snapshot file in the `root` directory
+   * and reports information about the process,
+   * including file size and timing metrics.
+   * Note that timing metrics exclude file open.
+   *
    * @param {(filePath: string) => Promise<void>} saveRaw
    * @returns {Promise<SnapshotInfo>}
    */
   async function save(saveRaw) {
-    return withTempName(async tmpSnapPath => {
-      const { duration: rawSaveSeconds } = await measureSeconds(() =>
-        saveRaw(tmpSnapPath),
-      );
-      const { size: rawByteCount } = await stat(tmpSnapPath);
-      const { result: hash, duration: hashSeconds } = await measureSeconds(() =>
-        fileHash(tmpSnapPath),
-      );
-      toDelete.delete(hash);
-      const baseName = baseNameFromHash(hash);
-      const filePath = resolve(root, baseName);
-      const infoBase = {
-        filePath,
-        hash,
-        rawByteCount,
-        rawSaveSeconds,
-        hashSeconds,
-      };
-      const fileStat = await stat(filePath).catch(e => {
-        if (e.code === 'ENOENT') {
-          return undefined;
+    const cleanup = [];
+    return aggregateTryFinally(
+      async () => {
+        // TODO: Refactor to use tmpFile rather than tmpName.
+        const tmpSnapPath = await ptmpName({ template: 'save-raw-XXXXXX.xss' });
+        cleanup.push(() => unlink(tmpSnapPath));
+        const { duration: rawSaveSeconds } = await measureSeconds(async () =>
+          saveRaw(tmpSnapPath),
+        );
+        const { size: rawByteCount } = await stat(tmpSnapPath);
+
+        // Perform operations that read snapshot data in parallel.
+        // We still serialize the stat and opening of tmpSnapPath
+        // and creation of tmpGzPath for readability, but we could
+        // parallelize those as well if the cost is significant.
+        const snapReader = createReadStream(tmpSnapPath);
+        cleanup.push(() => {
+          snapReader.destroy();
+        });
```

Comment:
Cleanup ignores return values and the other uses of `cleanup.push` don't suppress them like this.
```suggestion
        cleanup.push(() => snapReader.destroy());
```

Reply:
The return value would be adopted in the promise result in the `finally` stage, meaning the stream object would be kept around for longer than necessary.

Reply:
But just in the time between running the function in `PromiseAllOrErrors` and throwing away the aggregated value, which should be inconsequential.